test/integration/scheduler/priorities_test.go

```diff
@@ -0,0 +1,174 @@
+		Namespace: context.ns.Name,
+		Affinity: &v1.Affinity{
+			NodeAffinity: &v1.NodeAffinity{
+				PreferredDuringSchedulingIgnoredDuringExecution: []v1.PreferredSchedulingTerm{
+					{
+						Preference: v1.NodeSelectorTerm{
+							MatchExpressions: []v1.NodeSelectorRequirement{
+								{
+									Key:      labelKey,
+									Operator: v1.NodeSelectorOpIn,
+									Values:   []string{labelValue},
+								},
+							},
+						},
+						Weight: 20,
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("Error running pause pod: %v", err)
+	}
+	if pod.Spec.NodeName != labeledNode.Name {
+		t.Errorf("Pod %v got scheduled on an unexpected node: %v. Expected node: %v.", podName, pod.Spec.NodeName, labeledNode.Name)
+	} else {
+		t.Logf("Pod %v got successfully scheduled on node %v.", podName, pod.Spec.NodeName)
+	}
+}
+
+// TestPodAffinity verifies that scheduler's pod affinity priority function
+// works correctly.
+func TestPodAffinity(t *testing.T) {
+	context := initTest(t, "pod-affinity")
+	defer cleanupTest(t, context)
+	// Add a few nodes.
+	nodesInTopology, err := createNodes(context.clientSet, "in-topology", 5)
+	if err != nil {
+		t.Fatalf("Cannot create nodes: %v", err)
+	}
+	topologyKey := "node-topologykey"
+	topologyValue := "topologyvalue"
+	nodeLabels := map[string]string{
+		topologyKey: topologyValue,
+	}
+	for _, node := range nodesInTopology {
+		// Add topology key to all the nodes.
+		if err = testutils.AddLabelsToNode(context.clientSet, node.Name, nodeLabels); err != nil {
+			t.Fatalf("Cannot add labels to node %v: %v", node.Name, err)
+		}
+		if err = waitForNodeLabels(context.clientSet, node.Name, nodeLabels); err != nil {
+			t.Fatalf("Adding labels to node %v didn't succeed: %v", node.Name, err)
+		}
+	}
+	// Add a pod with a label and wait for it to schedule.
+	labelKey := "service"
+	labelValue := "S1"
+	_, err = runPausePod(context.clientSet, &pausePodConfig{
+		Name:      "attractor-pod",
+		Namespace: context.ns.Name,
+		Labels:    map[string]string{labelKey: labelValue},
+	})
+	if err != nil {
+		t.Fatalf("Error running the attractor pod: %v", err)
+	}
+	// Add a few more nodes without the topology label.
+	_, err = createNodes(context.clientSet, "other-node", 5)
+	if err != nil {
+		t.Fatalf("Cannot create the second set of nodes: %v", err)
+	}
+	// Add a new pod with affinity to the attractor pod.
+	podName := "pod-with-podaffinity"
+	pod, err := runPausePod(context.clientSet, &pausePodConfig{
+		Name:      podName,
+		Namespace: context.ns.Name,
+		Affinity: &v1.Affinity{
+			PodAffinity: &v1.PodAffinity{
+				PreferredDuringSchedulingIgnoredDuringExecution: []v1.WeightedPodAffinityTerm{
+					{
+						PodAffinityTerm: v1.PodAffinityTerm{
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      labelKey,
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{labelValue, "S3"},
```

Comment:
Shouldn't this be S1? It seems that this test is passing only by luck. Multiple MatchExpressions are ANDed, so the overall pod affinity term should be false on all nodes.